### PR TITLE
[WIP] Add KTOLoss

### PR DESCRIPTION
**Pull request by krammnic**
#### Context
What is the purpose of this PR? Is it to
- [X ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
#1850 
#### Changelog
What are the changes made in this PR?
* Adding KTO loss to torchtune

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [X] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [X] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [X] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [X] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by krammnic** [comment]
- [ ] Run some recipe with this loss.
- [ ] Add APO loss variant
- [ ] KL divergence?

**Comment by krammnic** [comment]
@SalmanMohammadi Should I make a full run with KTOLoss?

**Comment by SalmanMohammadi** [comment]
Hey @krammnic. Thanks so much for opening this. This is a great start.

I've been thinking about to improve the contributor workflow, and for contributions which may involve design decisions i.e. touch our existing codebase in a non-trivial way, require refactoring, or generally need some carefulness to make sure we're exposing new features to users in the best way possible.

So, before we dive into the specific implementation details, I'd like to take a step back. To prevent thrash and lots of back-and-forth and discovering design reconsiderations later into the PR, let's de-risk things by better understanding what we're doing here, and agree on some of the necessary steps we need to see this landed. I'm going to shamelessly plug my SimPO PR (https://github.com/pytorch/torchtune/pull/1223) as an example of this.

Would you be willing to answer some questions for me by updating the PR description?
- Could you provide a very very high-level overview of what this method is? What are the main ways it differs from e.g. DPO? You could provide some information from the paper here.
- Have you used a reference implementation? If so, which one? Code pointers here are very useful. I like to use TRL (https://github.com/huggingface/trl) which I believe has a KTOTrainer, but others also use official repos from paper authors.
- What would integrating this into the DPO recipes look like? Does it fit neatly into the existing DPO logic, or do we need some specific logic in the recipe for KTO? 
    - From skimming the KTOTrainer [docpage](https://huggingface.co/docs/trl/main/en/kto_trainer), it looks like TRL specify a different dataset format which differs from our preference dataset format in a meaningful way. As above with outlining the main differences between KTO and DPO, how would this different dataset format be used?

Generally what we're looking for is to make sure we don't find any gotchas down the line and that we can have an overview of the expected changes and outcomes for this contribution. 

A couple disclaimers:
- If this sounds like overkill - let me know! If you feel like there's a different way you'd like to go about it, please feel free! Similarly, I'm more than happy to help investigate on some of these points. You can also drop me a line on Discord : )
- Also, if this seems like a lot of effort, we can come back to this PR and you could try one of the RLHF techniques that you mentioned which may be easier to implement into our codebase. That's totally fine.

**Comment by krammnic** [comment]
Thanks for comment! I think there are already answers on most of your question in docstring of the KTOLoss) And in general RLHF issue. I might duplicate them here if required. 

Recipe should not be significantly changed. Change loss to KTOLoss. Will check about the dataset but generally in all contrastives they are pretty same

File: torchtune/rlhf/loss/dpo.py [hunk 1]
```diff
@@ -231,3 +231,91 @@ def forward(
...
+
+class KTOLoss(nn.Module):
+    """
+    KTO: Kahneman-Tversky Optimization: https://arxiv.org/abs/2402.01306
+    Intuition from the paper:
+
+        The effectiveness of SimPO is attributed to a key design: using the average log probability of a sequence as
+        the implicit reward. Additionally, we introduce a target reward margin to the Bradley-Terry objective to
+        encourage a larger margin between the winning and losing responses, further enhancing the algorithm's performance.
+
```
Comment (SalmanMohammadi) [inline torchtune/rlhf/loss/dpo.py:244]: It looks like this has been copied from the SimPO loss?

**Comment by krammnic** [comment]
We can conclude main idea of the method from its limitations! KTO >= DPO and DPO should be used in cases with high quality data, in other cases KTO is really suitable(Actually, there are 2 versions of this loss, which gives more control on situation with your dataset, but they are minorly different, so usually included in one loss(will do it as follow up))

File: torchtune/rlhf/loss/dpo.py (hunk 1 above)
Comment (krammnic) [reply torchtune/rlhf/loss/dpo.py:244]: Oh, I used SimPO docs as template and forgot to add my docs in this part. Other comments are about KTO

Comment (krammnic) [reply torchtune/rlhf/loss/dpo.py]: Fixed

File: torchtune/rlhf/loss/dpo.py [hunk 2]
```diff
@@ -231,3 +231,91 @@ def forward(
...
+        losses = torch.cat(
+            (
+                self.desirable_weight * chosen_losses,
+                self.undesirable_weight * rejected_losses,
+            ),
+            0,
+        )
+
+        chosen_rewards = self.beta * (chosen_logratios).detach()
+        rejected_rewards = self.beta * (rejected_logratios).detach()
```
Comment (SalmanMohammadi) [inline torchtune/rlhf/loss/dpo.py:319]: I'm not convinced that this is correct. Could you share a reference implementation for this loss function? For example, in TRL, the KL calculation is done differently (https://github.com/huggingface/trl/blob/main/trl/trainer/kto_trainer.py#L1100) - you can see that `self.calculate_KL` is `True` for the KTO loss (https://github.com/huggingface/trl/blob/main/trl/trainer/kto_trainer.py#L545).

**Comment by SalmanMohammadi** [comment]
> I might duplicate them here if required.

That would be very helpful to see! I'm also not convinced about the dataset format, as I've also found in the official KTO repo (https://github.com/ContextualAI/HALOs?tab=readme-ov-file) that:
> KTO doesn't use preference pairs, just knowledge of whether outputs are desirable or undesirable. This means we use dataloader.UnpairedPreferenceDataLoader. However, that dataloader assumes that you're working with datasets that originally contain preference pairs, like Anthropic HH or SHP
So we need to think a bit about how to make sure our implementation in the DPO recipe retains parity here. We could use some minor modification of a preference dataset here, but I'd like to make sure we're aware of these kinds of difference.

File: torchtune/rlhf/loss/dpo.py (hunk 2 above)
Comment (SalmanMohammadi) [reply torchtune/rlhf/loss/dpo.py:319]: This is the same in the official KTO repo https://github.com/ContextualAI/HALOs/blob/61b9ee6787c9f52c2f4578533866e1cb42f314ec/trainers.py#L836

Comment (krammnic) [reply torchtune/rlhf/loss/dpo.py:319]: Yes, I saved this mostly as drafted(version without KL) will bring it

**Comment by krammnic** [comment]
> > I might duplicate them here if required.
> 
> That would be very helpful to see! I'm also not convinced about the dataset format, as I've also found in the official KTO repo (https://github.com/ContextualAI/HALOs?tab=readme-ov-file) that:
> 
> > KTO doesn't use preference pairs, just knowledge of whether outputs are desirable or undesirable. This means we use dataloader.UnpairedPreferenceDataLoader. However, that dataloader assumes that you're working with datasets that originally contain preference pairs, like Anthropic HH or SHP
> 
> So we need to think a bit about how to make sure this implementation in the DPO recipe retains parity here. Do we need use some minor modification of a preference dataset here?

Yes, probably will require special things about data loading

File: torchtune/rlhf/loss/dpo.py (hunk 2 above)
Comment (krammnic) [reply torchtune/rlhf/loss/dpo.py:319]: Hmm, to actually get kl_logprobs we need to modify concatenated_forward

Comment (krammnic) [reply torchtune/rlhf/loss/dpo.py:319]: I don't want to pass cfg in concatenated_forward

Comment (krammnic) [reply torchtune/rlhf/loss/dpo.py:319]: We can actually ignore KL divergence and use APO variant of loss:

`rejected_losses = F.sigmoid(self.beta * rejected_logratios)`

Comment (krammnic) [reply torchtune/rlhf/loss/dpo.py:319]: And we need to actually change structure of the batch

**Comment by krammnic** [comment]
> > > I might duplicate them here if required.
> > 
> > 
> > That would be very helpful to see! I'm also not convinced about the dataset format, as I've also found in the official KTO repo (https://github.com/ContextualAI/HALOs?tab=readme-ov-file) that:
> > > KTO doesn't use preference pairs, just knowledge of whether outputs are desirable or undesirable. This means we use dataloader.UnpairedPreferenceDataLoader. However, that dataloader assumes that you're working with datasets that originally contain preference pairs, like Anthropic HH or SHP
> > 
> > 
> > So we need to think a bit about how to make sure this implementation in the DPO recipe retains parity here. Do we need use some minor modification of a preference dataset here?
> 
> Yes, probably will require special things about data loading

Looks like it want be difficult, but actually we can create new dataset type for this. So basically to convert pair dataset to KTO format we need something like:

```
batch_size = len(examples["chosen"])
new_rows = {
        "completion": examples["chosen"] + examples["rejected"],
        "label": [True] * batch_size + [False] * batch_size,
 }
if "prompt" in examples:
        new_rows["prompt"] = examples["prompt"] + examples["prompt"]
```

And obviously remove required columns. I'm not sure where to put such preprocessing. Either new dataset type, or new argument in current dataset?

**Comment by SalmanMohammadi** [comment]
> And we need to actually change structure of the batch. Also would require some processing to dataset. Do you approve such changes?

Would you be able to outline some of the necessary steps we need to make on the dataset side?

> We can actually ignore KL divergence right now and use APO variant of loss:
  rejected_losses = F.sigmoid(self.beta * rejected_logratios)

APO is a separate (but comparable) technique to KTO, right? Is [this](https://huggingface.co/papers/2408.06266) the right paper? It looks like it came out < 2 months ago, and I see very few models on the hub mentioning APO. As I mentioned in the RLHF tracker, I'm not sure we should be implementing techniques which are new and haven't been adopted by the OS community, so I'd rather we stick with the original proposal of KTO, and I'd love to keep supporting you on this. 

I don't fully follow your comments on the necessary modifications needed to implement KTO, you mentioned needing to modify `concatenated_forward`? I'm not sure I fully follow along so it'd be really helpful if you could also outline why we need to make these changes - that way we can figure out the best way to implement them.

**Comment by krammnic** [comment]
1. About APO. It looks like it is just KTO without KL, but ok, let's skip
2. Therefore, we don't need to pass anything extra in forward!
3. We need to do this preprocessing to dataset:
```
batch_size = len(examples["chosen"])
new_rows = {
        "completion": examples["chosen"] + examples["rejected"],
        "label": [True] * batch_size + [False] * batch_size,
 }
if "prompt" in examples:
        new_rows["prompt"] = examples["prompt"] + examples["prompt"]
``` and map it doataset

4. But we need to calculate: `policy_KL_logps, reference_KL_logps`
5. then we need batch["reference_KL_logps"] and policy_KL_logps

**Comment by krammnic** [comment]
> 1. About APO. It looks like it is just KTO without KL, but ok, let's skip
> 2. Therefore, we don't need to pass anything extra in forward!
> 3. We need to do this preprocessing to dataset:
> 
> ```
> batch_size = len(examples["chosen"])
> new_rows = {
>         "completion": examples["chosen"] + examples["rejected"],
>         "label": [True] * batch_size + [False] * batch_size,
>  }
> if "prompt" in examples:
>         new_rows["prompt"] = examples["prompt"] + examples["prompt"]
> ```
> 
> and map it to dataset
> 
> 4. But we need to calculate: `policy_KL_logps, reference_KL_logps`
> 5. then we need batch["reference_KL_logps"] and policy_KL_logps

@SalmanMohammadi Can we continue from here?

**Comment by SalmanMohammadi** [comment]
> > 1. About APO. It looks like it is just KTO without KL, but ok, let's skip
> > 2. Therefore, we don't need to pass anything extra in forward!
> > 3. We need to do this preprocessing to dataset:
> > 
> > ```
> > batch_size = len(examples["chosen"])
> > new_rows = {
> >         "completion": examples["chosen"] + examples["rejected"],
> >         "label": [True] * batch_size + [False] * batch_size,
> >  }
> > if "prompt" in examples:
> >         new_rows["prompt"] = examples["prompt"] + examples["prompt"]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and map it to dataset
> > 
> > 4. But we need to calculate: `policy_KL_logps, reference_KL_logps`
> > 5. then we need batch["reference_KL_logps"] and policy_KL_logps
> 
> @SalmanMohammadi Can we continue from here?

Hey @krammnic - thanks for summarizing here, it seems like we can make things simpler if we don't try to add APO too? 

I need a bit of time to look into your other questions which I may not get round to until later this week. If you have an idea of how you'd like to proceed here I'd say go for it and we can try iterate if we need to. I'm interested in your thoughts on whether you think we'll need to make any changes to the recipe itself - for example, does `concatenated_forward` need to be changed? Or can we make modifications just at the dataset level?

Out of curiosity, where did you grab this from?

> batch_size = len(examples["chosen"])
new_rows = {
        "completion": examples["chosen"] + examples["rejected"],
        "label": [True] * batch_size + [False] * batch_size,
 }
if "prompt" in examples:
        new_rows["prompt"] = examples["prompt"] + examples["prompt"]

**Comment by krammnic** [comment]
Thanks for the answer!

Yes some changes will be required as we need to get batch["reference_KL_logps"] and policy_KL_logps to calculate KL divergence(will write what we need to do here soon). 

Dataset processing implementation is from TRL:

https://github.com/huggingface/trl/blob/main/trl/data_utils.py

**Comment by krammnic** [comment]
Added dataset processing, now we need to add correct KL divergence calculation

**Comment by krammnic** [comment]
@SalmanMohammadi Are we supporting different ref_model in current DPO recipe? As I see, we are not: 

```
with torch.no_grad(), disable_adapter(self._model):
      (
          reference_chosen_log_probs,
          reference_rejected_log_probs,
                            _,
                            _,
      ) = self.concatenated_forward(self._model, batch)
```

**Comment by SalmanMohammadi** [comment]
> @SalmanMohammadi Are we supporting different ref_model in current DPO recipe? As I see, we are not:
> 
> ```
> with torch.no_grad(), disable_adapter(self._model):
>       (
>           reference_chosen_log_probs,
>           reference_rejected_log_probs,
>                             _,
>                             _,
>       ) = self.concatenated_forward(self._model, batch)
> ```

What do you mean by a different ref model? Because we're fine-tuning with LoRA, we can just disable the LoRA adapters to recover the original reference model without having to keep a separate copy of the model in memory. The reference model should always be the base model we're finetuning from.

**Comment by krammnic** [comment]
> > @SalmanMohammadi Are we supporting different ref_model in current DPO recipe? As I see, we are not:
> > ```
> > with torch.no_grad(), disable_adapter(self._model):
> >       (
> >           reference_chosen_log_probs,
> >           reference_rejected_log_probs,
> >                             _,
> >                             _,
> >       ) = self.concatenated_forward(self._model, batch)
> > ```
> 
> What do you mean by a different ref model? Because we're fine-tuning with LoRA, we can just disable the LoRA adapters to recover the original reference model without having to keep a separate copy of the model in memory. The reference model should always be the base model we're finetuning from.

Actually in TRL there is possibility to set something different! 

 ```
ref_model (`PreTrainedModelWrapper`):
            Hugging Face transformer model with a casual language modelling head. Used for implicit reward computation and loss. If no
            reference model is provided, the trainer will create a reference model with the same architecture as the model to be optimized.
``` 

I'm not sure why they give this possibility either, so that why I've asked. Maybe there is special use case?

**Comment by SalmanMohammadi** [comment]
Ah good catch! I'm not 100% sure either. In their example scripts they always initialize with the same model (https://github.com/huggingface/trl/blob/main/examples/scripts/kto.py).

**Comment by krammnic** [comment]
Actually, it is pretty critical to DPO, for example according to https://arxiv.org/pdf/2407.13709

```
Our experiments reveal an interesting finding: stronger
reference models can indeed offer more benefits
than the SFT model, but only when they are compat-
ible with the model being fine-tuned
```

I'm pretty sure that it will be actual for classic DPO for example

**Comment by krammnic** [comment]
I would consider this as a follow up, which we might need to add pretty soon

**Comment by krammnic** [comment]
Tasks:

- [ ] Fix test
- [ ] Modify batch
- [ ] Do same changes to distributed recipe
